File: document-ui/src/main/java/dev/sayaya/handbook/client/domain/Type.java
package dev.sayaya.handbook.client.domain;

import lombok.Builder;
import lombok.Data;
import lombok.Singular;
import lombok.experimental.Accessors;

import java.util.Date;
import java.util.List;
import java.util.Objects;

@Data
@Accessors(fluent = true)
@Builder(toBuilder = true)
public class Type {
    private final String id;
    private final String version;
    private final String prev;
    private final String next;
    private final Date effectDateTime;
    private final Date expireDateTime;
    private final String description;
    private final boolean primitive;
    @Singular("attribute")
    private final List<Attribute> attributes;
    private final String parent;

    private Type(String id, String version, String prev, String next,
                 Date effectDateTime, Date expireDateTime, String description,
                 boolean primitive, List<Attribute> attributes, String parent) {
        this.id = validateNonNullOrEmpty(id, "id");
        this.version = validateNonNullOrEmpty(version, "version");
        this.prev = prev;
        this.next = next;
        this.effectDateTime = validateNonNull(effectDateTime, "Effect date time");
        this.expireDateTime = validateNonNull(expireDateTime, "Expire date time");
        this.description = description;
        this.primitive = primitive;
        this.attributes = validateNonNull(attributes, "attributes");
        this.parent = parent;
        require(expireDateTime.after(effectDateTime), "Expire date time must be after effect date time");
    }
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Type type = (Type) o;
        return Objects.equals(id, type.id) && Objects.equals(version, type.version);
    }
    @Override
    public int hashCode() {
        return Objects.hash(id, version);
    }

    private static void require(boolean condition, String message) {
        if (!condition) throw new IllegalArgumentException(message);
    }
    private static <T> T validateNonNull(T obj, String message) {
        require(obj != null, message + " must not be null");
        return obj;
    }
    private static String validateNonNullOrEmpty(String obj, String message) {
        require(obj != null && !obj.isEmpty(), message + " must not be null");
        return obj;
    }
}

File: domain/src/main/kotlin/dev/sayaya/handbook/domain/Type.kt
package dev.sayaya.handbook.domain

import java.io.Serializable
import java.time.Instant
import java.util.Objects

/**
 * 시스템에서 사용되는 데이터의 '타입'을 정의하는 엔티티(Entity).
 *
 * 이 엔티티의 고유 식별자는 사람이 읽을 수 있는 `id`와 `version`의 조합(복합키)입니다.
 * 예를 들어, 'post-v1'과 'post-v2'는 서로 다른 타입입니다.
 */
@JvmRecord
data class Type (
    val id: String,
    val version: String,
    val effectDateTime: Instant,
    val expireDateTime: Instant,
    val description: String?,
    val primitive: Boolean,
    // val attributes: List<Attribute> = emptyList(),
    // 부모 타입의 ID를 값으로 직접 참조하여 다른 Aggregate 간의 느슨한 결합을 유지합니다.
    val parent: String? = null,

    ): Serializable {
    init {
        require(id.isNotBlank()) { "Type id cannot be blank" }
        require(id.matches(Regex("^[a-zA-Z0-9가-힣_-]+$"))) { "Type id can only contain alphabet, 한글, numbers, hyphens, and underscores." }
        require(expireDateTime.isAfter(effectDateTime)) { "Expire date time must be after effect date time" }
    }

    /**
     * 데이터 클래스의 기본 동작을 따르지 않고,
     * 복합키(`id`, `version`) 기반의 동등성 규칙을 적용하기 위해 재정의합니다.
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Type
        return id == other.id && version == other.version
    }

    /**
     * equals가 복합키를 사용하므로, hashCode도 복합키를 기반으로 생성합니다.
     */
    override fun hashCode(): Int {
        return Objects.hash(id, version)
    }
}